cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -837,6 +834,54 @@ private void setOutputsIfNeeded(Job job) {
     }
   }
 
+  private void fixOutputPermissions(Job job, Map<String, ProvidedOutput> outputs) {
+    Set<String> outputsWithUmask = new HashSet<>();
+    Set<String> outputUmasks = new HashSet<>();
+    for (Map.Entry<String, ProvidedOutput> entry : outputs.entrySet()) {
+      String umask = entry.getValue().getOutputFormatConfiguration().get("fs.permissions.umask-mode");
+      if (umask != null) {
+        outputsWithUmask.add(entry.getKey());
+        outputUmasks.add(umask);
+      }
+    }
+    boolean allOutputsHaveUmask = outputsWithUmask.size() == outputs.size();
+    boolean allOutputsAgree = outputUmasks.size() == 1;
+    boolean jobConfHasUmask = ConfigurationUtil.isProgrammaticProperty(
+      job.getConfiguration().getPropertySources("fs.permissions.umask-mode"));
+    String jobConfUmask = job.getConfiguration().get("fs.permissions.umask-mode");
+
+    boolean mustFixUmasks = false;
+    if (jobConfHasUmask) {
+      // case 1: job conf has a programmatic umask. It prevails.
+      mustFixUmasks = !outputsWithUmask.isEmpty();
+      LOG.info("Overriding permissions of outputs {} because a umask of {} was set programmatically in the job " +
+                 "configuration.", outputsWithUmask, jobConfUmask);
+    } else if (allOutputsHaveUmask && allOutputsAgree) {
+      // case 2: no programmatic umask in job conf, all outputs want the same umask: set it in job conf
+      String umaskToUse = outputUmasks.iterator().next();
+      job.getConfiguration().set("fs.permissions.umask-mode", umaskToUse);
+      LOG.debug("Setting umask of {} in job configuration because all outputs {} agree on it.",
+                umaskToUse, outputsWithUmask);
+    } else {
+      // case 3: some outputs configure a umask, but not all of them, or not all the same: use job conf default
+      mustFixUmasks = !outputsWithUmask.isEmpty();
+      LOG.warn("Overriding permissions of outputs {} because they configure different permissions. Falling back to " +
```

Comment:
We only need to log this if `mustFixUmasks` is `true`, right? Because if none of the output has umask set, we don't need to fix the umask, hence no overriding of permissions due to disagreement.